Mathlib/Topology/MetricSpace/HolderNorm.lean

```diff
@@ -0,0 +1,230 @@
+
+variable (X) in
+lemma eHolderNorm_zero [Zero Y] (r : ℝ≥0) : eHolderNorm r (0 : X → Y) = 0 :=
+  eHolderNorm_const X r 0
+
+attribute [simp] eHolderNorm_const eHolderNorm_zero
+
+lemma eHolderNorm_of_isEmpty [hX : IsEmpty X] :
+    eHolderNorm r f = 0 := by
+  rw [eHolderNorm, ← ENNReal.bot_eq_zero, iInf₂_eq_bot]
+  exact fun ε hε => ⟨0, .of_isEmpty, hε⟩
+
+lemma HolderWith.eHolderNorm_le {C : ℝ≥0} (hf : HolderWith C r f) :
+    eHolderNorm r f ≤ C :=
+  iInf₂_le C hf
+
+@[simp]
+lemma memHolder_const {c : Y} : MemHolder r (Function.const X c) :=
+  (HolderWith.const (C := 0)).memHolder
+
+@[simp]
+lemma memHolder_zero [Zero Y] : MemHolder r (0 : X → Y) :=
+  memHolder_const
+
+end PseudoEMetricSpace
+
+section MetricSpace
+
+variable [MetricSpace X] [EMetricSpace Y]
+
+lemma eHolderNorm_eq_zero {r : ℝ≥0} {f : X → Y} :
+    eHolderNorm r f = 0 ↔ ∀ x₁ x₂, f x₁ = f x₂ := by
+  constructor
+  · refine fun h x₁ x₂ => ?_
+    by_cases hx : x₁ = x₂
+    · rw [hx]
+    · rw [eHolderNorm, ← ENNReal.bot_eq_zero, iInf₂_eq_bot] at h
+      rw [← edist_eq_zero, ← le_zero_iff]
+      refine le_of_forall_lt' fun b hb => ?_
+      obtain ⟨C, hC, hC'⟩ := h (b / edist x₁ x₂ ^ (r : ℝ))
+        (ENNReal.div_pos hb.ne.symm (ENNReal.rpow_lt_top_of_nonneg zero_le_coe
+          (edist_lt_top x₁ x₂).ne).ne)
+      exact lt_of_le_of_lt (hC x₁ x₂) <| ENNReal.mul_lt_of_lt_div hC'
+  · intro h
+    cases' isEmpty_or_nonempty X with hX hX
+    · haveI := hX
+      exact eHolderNorm_of_isEmpty
+    · rw [← eHolderNorm_const X r (f hX.some)]
+      congr
+      simp [funext_iff, h _ hX.some]
+
+lemma MemHolder.holderWith {r : ℝ≥0} {f : X → Y} (hf : MemHolder r f) :
+    HolderWith (nnHolderNorm r f) r f := by
+  intros x₁ x₂
+  by_cases hx : x₁ = x₂
+  · simp only [hx, edist_self, zero_le]
+  rw [nnHolderNorm, eHolderNorm, coe_toNNReal]
+  swap; exact hf.eHolderNorm_lt_top.ne
+  have h₁ : edist x₁ x₂ ^ (r : ℝ) ≠ 0 :=
+    (Ne.symm <| ne_of_lt <| ENNReal.rpow_pos (edist_pos.2 hx) (edist_lt_top x₁ x₂).ne)
+  have h₂ : edist x₁ x₂ ^ (r : ℝ) ≠ ∞ := by
+    simp [(edist_lt_top x₁ x₂).ne]
+  rw [← ENNReal.div_le_iff h₁ h₂]
+  refine le_iInf₂ fun C hC => ?_
+  rw [ENNReal.div_le_iff h₁ h₂]
+  exact hC x₁ x₂
+
+lemma memHolder_iff_holderWith {r : ℝ≥0} {f : X → Y} :
+    MemHolder r f ↔ HolderWith (nnHolderNorm r f) r f :=
+  ⟨MemHolder.holderWith, HolderWith.memHolder⟩
+
+lemma coe_nnHolderNorm_le_eHolderNorm {r : ℝ≥0} {f : X → Y} :
+    (nnHolderNorm r f : ℝ≥0∞) ≤ eHolderNorm r f :=
+  coe_toNNReal_le_self
+
+lemma HolderWith.coe_nnHolderNorm_eq_eHolderNorm
+    {C r : ℝ≥0} {f : X → Y} (hf : HolderWith C r f) :
+    (nnHolderNorm r f : ℝ≥0∞) = eHolderNorm r f := by
+  rw [nnHolderNorm, coe_toNNReal]
+  exact ne_of_lt <| lt_of_le_of_lt hf.eHolderNorm_le <| coe_lt_top (r := C)
+
+lemma MemHolder.coe_nnHolderNorm_eq_eHolderNorm
+    {r : ℝ≥0} {f : X → Y} (hf : MemHolder r f) :
+    (nnHolderNorm r f : ℝ≥0∞) = eHolderNorm r f :=
+  hf.holderWith.coe_nnHolderNorm_eq_eHolderNorm
```

Comment:
I would keep only the second of these two given that the conclusion doesn't mention `C`

Reply:
I don't see whats the harm in keeping it.

Reply:
API streamlining 🤷